File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/rules/PelagoPushRouterDown.java
package ch.epfl.dias.calcite.adapter.pelago.rules;

import org.apache.calcite.plan.RelOptRule;
import org.apache.calcite.plan.RelOptRuleCall;
import org.apache.calcite.rel.RelNode;

import ch.epfl.dias.calcite.adapter.pelago.PelagoFilter;
import ch.epfl.dias.calcite.adapter.pelago.PelagoProject;
import ch.epfl.dias.calcite.adapter.pelago.PelagoRouter;
import ch.epfl.dias.calcite.adapter.pelago.RelDeviceType;
import ch.epfl.dias.calcite.adapter.pelago.RelDeviceTypeTraitDef;
import ch.epfl.dias.calcite.adapter.pelago.RelHomDistribution;

import java.util.stream.Collectors;

public class PelagoPushRouterDown extends RelOptRule {

  public static final RelOptRule[] RULES = new RelOptRule[]{
    new PelagoPushRouterDown(PelagoFilter.class, RelHomDistribution.RANDOM),
    new PelagoPushRouterDown(PelagoProject.class, RelHomDistribution.RANDOM),
    new PelagoPushRouterDown(PelagoFilter.class, RelHomDistribution.BRDCST),
    new PelagoPushRouterDown(PelagoProject.class, RelHomDistribution.BRDCST),
    PelagoPushRouterBelowJoin.INSTANCE,
    PelagoPushRouterBelowAggregate.INSTANCE, //NoPartial
  };

  protected final RelHomDistribution trgt;

  protected PelagoPushRouterDown(Class<? extends RelNode> op,
      final RelHomDistribution trgt) {
    super(operand(op, any()), "PPRD" + op.getName() + trgt.toString());
//    this.op = op;
    this.trgt = trgt;
  }

  protected static RelNode route(RelNode input, RelHomDistribution trgt){
    return convert(
      input,
      input.getTraitSet().replace(trgt).replaceIf(RelDeviceTypeTraitDef.INSTANCE, () -> RelDeviceType.X86_64)
    );
  }

  protected RelNode route(RelNode input){
    return route(input, trgt);
  }

  public boolean matches(RelOptRuleCall call){
    var trait = call.rel(0).getTraitSet();
    return trait.containsIfApplicable(RelHomDistribution.SINGLE);
  }

  public void onMatch(RelOptRuleCall call) {
    var rel = call.rel(0);

    call.transformTo(
      rel.copy(null,
        rel.getInputs().stream().map(this::route).collect(Collectors.toList())
      )
    );
  }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/rules/PelagoPushRouterDown.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */
package ch.epfl.dias.calcite.adapter.pelago.rules

import ch.epfl.dias.calcite.adapter.pelago.rel.{PelagoFilter, PelagoProject}
import ch.epfl.dias.calcite.adapter.pelago.traits.{
  RelDeviceType,
  RelDeviceTypeTraitDef,
  RelHomDistribution
}
import org.apache.calcite.plan.RelOptRule.{any, convert, operand}
import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
import org.apache.calcite.rel.RelNode

import scala.collection.JavaConverters._

object PelagoPushRouterDown {
  val RULES: Array[RelOptRule] = Array[RelOptRule](
    new PelagoPushRouterDown(classOf[PelagoFilter], RelHomDistribution.RANDOM),
    new PelagoPushRouterDown(classOf[PelagoProject], RelHomDistribution.RANDOM),
    new PelagoPushRouterDown(classOf[PelagoFilter], RelHomDistribution.BRDCST),
    new PelagoPushRouterDown(classOf[PelagoProject], RelHomDistribution.BRDCST),
    PelagoPushRouterBelowJoin.INSTANCE,
    PelagoPushRouterBelowAggregate.INSTANCE
  )

  def route(
      input: RelNode,
      trgt: RelHomDistribution
  ): RelNode =
    convert(
      input,
      input.getTraitSet
        .replace(trgt)
        .replaceIf(RelDeviceTypeTraitDef.INSTANCE, () => RelDeviceType.X86_64)
    )
}

class PelagoPushRouterDown protected (
    val op: Class[_ <: RelNode],
    val trgt: RelHomDistribution
) extends RelOptRule(operand(op, any), "PPRD" + op.getName + trgt.toString) {

  protected def route(input: RelNode): RelNode =
    PelagoPushRouterDown.route(input, trgt)

  override def matches(call: RelOptRuleCall): Boolean = {
    call
      .rel(0)
      .asInstanceOf[RelNode]
      .getTraitSet
      .containsIfApplicable(RelHomDistribution.SINGLE)
  }

  override def onMatch(call: RelOptRuleCall): Unit = {
    val rel = call.rel(0).asInstanceOf[RelNode]
    call.transformTo(
      rel.copy(
        null,
        rel.getInputs.asScala.map(this.route).asJava
      )
    )
  }
}
